### PR TITLE
add channels last with mixed data type support for GroupNorm backward

### PR DESCRIPTION
**Pull request by CaoE**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #89485

### Motivation
1. Add channels last support for GroupNorm backward to make sure GroupNorm fully support channels last.
2. Same as #88663, mixed data type support is also needed for channels last implementation of GroupNorm backward.

### Testing
Single socket (28cores):

* Contiguous:

shape | forward / s | forward / s | backward / s | backward / s
-- | -- | -- | -- | --
  | fp32 | mixed fp32 bf16 | fp32 | mixed fp32 bf16
[10, 128, 20, 20] | 3.20E-05 | 3.60E-05 | 8.31E-05 | 8.13E-05
[10, 128, 50, 50] | 0.000126 | 0.000115 | 0.000356 | 0.000257

* Channels Last:

shape | forward / s | forward / s | backward / s | backward / s
-- | -- | -- | -- | --
  | fp32 | mixed fp32 bf16 | fp32 | mixed fp32 bf16
[10, 128, 20, 20] | 4.11E-05 | 4.12E-05 | 9.74E-05 | 9.66E-05
[10, 128, 50, 50] | 0.000179 | 0.000178 | 0.000393 | 0.000317

Single core:

* Contiguous:

shape | forward / s | forward / s | backward / s | backward / s
-- | -- | -- | -- | --
  | fp32 | mixed fp32 bf16 | fp32 | mixed fp32 bf16
[10, 128, 20, 20] | 2.47E-04 | 2.53E-04 | 5.92E-04 | 4.50E-04
[10, 128, 50, 50] | 0.001559 | 0.001384 | 0.004343 | 0.002436

* Channels Last:

shape | forward / s | forward / s | backward / s | backward / s
-- | -- | -- | -- | --
  | fp32 | mixed fp32 bf16 | fp32 | mixed fp32 bf16
[10, 128, 20, 20] | 2.27E-04 | 3.24E-04 | 0.0006224 | 0.000459
[10, 128, 50, 50] | 0.00167 | 0.001278 | 0.0041858 | 0.003027

cc @ezyang @gchanan @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @VitalyFedyunin

### CODE REVIEW

**Review by jgong5 — APPROVED** [review]

File: aten/src/ATen/native/cpu/group_norm_kernel.cpp [hunk 1]
```diff
@@ -798,21 +1323,43 @@ void GroupNormBackwardKernelImpl(
...
+    case at::MemoryFormat::Contiguous: {
+      AT_DISPATCH_FLOATING_TYPES_AND(
+        ScalarType::BFloat16, X.scalar_type(), "GroupNormBackwardKernelImpl", [&]() {
+        if(mixed_type) {
+          GroupNormBackwardKernelImplInternal<BFloat16, float>(
+              dY, X, mean, rstd, gamma, N, C, HxW, group, dX, dgamma, dbeta);
+        } else {
+          GroupNormBackwardKernelImplInternal<scalar_t, scalar_t>(
+              dY, X, mean, rstd, gamma, N, C, HxW, group, dX, dgamma, dbeta);
+        }
```
Comment (malfet) [inline aten/src/ATen/native/cpu/group_norm_kernel.cpp:1341]: Same feedback as for other https://github.com/pytorch/pytorch/pull/81852 : can we somehow restrict this if to BFloat16 only (unless we can have a mixed precision case for float and double)

File: aten/src/ATen/native/cpu/group_norm_kernel.cpp [hunk 2]
```diff
@@ -785,6 +785,531 @@ void GroupNormBackwardKernelImplInternal(
...
+      }
+    }
+
+    // Step 3. Compute dx.
+    if (dX_data != nullptr) {
+      at::parallel_for(0, N * HxW, 1, [&](int64_t begin, int64_t end) {
+        int64_t n{0}, m{0};
+        data_index_init(begin, n, N, m, HxW);
+        for (const auto i : c10::irange(begin, end)) {
+          for (const auto g : c10::irange(0, G)) {
```
Comment (malfet) [inline aten/src/ATen/native/cpu/group_norm_kernel.cpp:1280]: Nit
```suggestion
          for (const auto g : c10::irange(G)) {
```

File: aten/src/ATen/native/cpu/group_norm_kernel.cpp [hunk 3]
```diff
@@ -785,6 +785,531 @@ void GroupNormBackwardKernelImplInternal(
...
+    // get the final internal gradients to ds, db, and tmp_buffer.
+    for (const auto n : c10::irange(N)) {
+      for (const auto g : c10::irange(G)) {
+        PT ds_gamma{0}, db_gamma{0};
+        for (const auto d : c10::irange(D)) {
+          PT ds_val{0}, db_val{0};
+          for (const auto t : c10::irange(num_threads)) {
+            PT* buffer_ptr = buffer_data + t * N * 2 * C + n * 2 * C;
+            ds_gamma += buffer_ptr[g * D + d] * (gamma_null ? PT(1) : gamma_data[g * D + d]);
+            db_gamma += buffer_ptr[g * D + d + C] * (gamma_null ? PT(1) : gamma_data[g * D + d]);
```
Comment (malfet) [inline aten/src/ATen/native/cpu/group_norm_kernel.cpp:1261]: Nit (compiler should be smart enough to optimize those out, but makes code a bit more readable.)
```suggestion
             if (gamma_null) {
                         ds_gamma += buffer_ptr[g * D + d];
                         db_gamma += buffer_ptr[g * D + d + C];
             } else {
                ds_gamma += buffer_ptr[g * D + d] *  gamma_data[g * D + d];
                db_gamma += buffer_ptr[g * D + d + C] * gamma_data[g * D + d];             
             }
```

**Review by malfet — APPROVED** [review]

File: aten/src/ATen/native/cpu/group_norm_kernel.cpp (hunk 1 above)
Comment (CaoE) [reply aten/src/ATen/native/cpu/group_norm_kernel.cpp]: This is bfloat16 only now, but float16 might be also added in the future. Use param_t = at::opmath_type<scalar_t> and GroupNormBackwardKernelImplInternal<scalar_t, param_t>for further  fp16 implementation.

File: aten/src/ATen/native/cpu/group_norm_kernel.cpp (hunk 2 above)
Comment (CaoE) [reply aten/src/ATen/native/cpu/group_norm_kernel.cpp]: Fixed it.

File: aten/src/ATen/native/cpu/group_norm_kernel.cpp (hunk 3 above)
Comment (CaoE) [reply aten/src/ATen/native/cpu/group_norm_kernel.cpp]: Fixed it.

**Comment by CaoE** [comment]
@malfet Is there any adjustments that can be made in later PRs. Thanks for your any suggestions.

**Comment by jgong5** [comment]
> @malfet Is there any adjustments that can be made in later PRs. Thanks for your any suggestions.

@malfet In particular, may I know what code duplicate you were concerned about in this PR. We can address this in the follow-up PR. Thanks.

**Comment by malfet** [comment]
This PR regressed training step for https://github.com/lucidrains/DALLE2-pytorch, see https://github.com/pytorch/pytorch/issues/92166

**Comment by CaoE** [comment]
@malfet  Sorry for the incontinence. Can we move the `.grads[0].contiguous` and `input.contiguous`  from https://github.com/pytorch/pytorch/pull/89485/files#diff-e4c2f99a2404e98c3586e07425da73008f36b1bada790648a7297af141d37f8cL1171 to the native_group_norm_backward function in aten/src/ATen/native/group_norm.cpp and add memory format check for X and dY . In this way, we can neither break the channel last on  CPU nor affect CUDA.

**Comment by malfet** [comment]
> @malfet Sorry for the incontinence. Can we move the `.grads[0].contiguous` and `input.contiguous` from https://github.com/pytorch/pytorch/pull/89485/files#diff-e4c2f99a2404e98c3586e07425da73008f36b1bada790648a7297af141d37f8cL1171 to the native_group_norm_backward function in aten/src/ATen/native/group_norm.cpp and add memory format check for X and dY . In this way, we can neither break the channel last on CPU nor affect CUDA.

That's what I'm doing: forcing the continuous on the CUDA side, but also adding the unit test so it would not regress again

**Comment by CaoE** [comment]
@malfet  Thank you very much ! Can I do anything to help fix this problem. I can submit a PR for the issue.

**Comment by CaoE** [comment]
@malfet I submitted a PR https://github.com/pytorch/pytorch/pull/92668 to fix it but I have no CUDA machine to verify if the issue is fixed. Could you please help to check it ?

**Comment by malfet** [comment]
@CaoE This sound very similar to the approach I've taken in https://github.com/pytorch/pytorch/pull/92671 and it fixes the problem for CUDA, but surprisingly I had to skip part of the test on CPU, as I think current implementation silently produces wrong results, but probably requires some further refactoring...

**Pull request closed without merging**